### ISSUE TITLE
[1.9.4] Email is truncated due to special chars

### ISSUE THREAD

**Issue by mtramp** [osTicket/osTicket#1475]
sticket version 1.9.4

Problem: An incoming email is truncated since it contains special chars.

I have verfied this issue with this test email:

Original email with header:

```
From - Tue Nov 04 08:48:05 2014
X-Account-Key: account7
X-UIDL: UI
X-Mozilla-Status: 0000
X-Mozilla-Status2: 00000000
X-Mozilla-Keys:                                                                                 
Return-Path: <originalemail@example.org>
X-Original-To: ouremail@example.org
Delivered-To: ouremail@example.org
Received: [...]
Subject: Verifiy osticket issue
From: =?utf-8?Q?Tester?= <myemail@example.org>
Content-Type: text/plain;
    charset=utf-8
X-Mailer: iPhone Mail (12A405)
Message-Id: <id@example.org>
Date: Mon, 3 Nov 2014 21:43:18 +0100
To: "department@example.org" <department@example.org>
Content-Transfer-Encoding: quoted-printable
Mime-Version: 1.0 (1.0)
X-Provags-ID:  [...]
X-UI-Out-Filterresults: notjunk:1;

Hello,
this is just a test mail with a special char.=20=

Do you see a smile at the end of "this"=F0=9F=98=83 string after
```

Osticket truncate the email text after "this". So I guess
=F0=9F=98=83
cause this trouble.

**Comment by blastoma** [comment]
@greezybacon hi! I have the same problem with v1.10-rc.2
But I have not tabe ost_ticket_thread. Can you help me?

**Comment by ntozier** [comment]
1.10rc2 
\upload\setup\inc\streams\core\install-mysql.sql 
does not have the SQL query to build the %ticket_thread table.
**but**
1.9.12
\upload\setup\inc\streams\core\install-mysql.sql
does!

Run this query to get the table:
DROP TABLE IF EXISTS `%TABLE_PREFIX%ticket_thread`;
CREATE TABLE `%TABLE_PREFIX%ticket_thread` (
  `id` int(11) unsigned NOT NULL auto_increment,
  `pid` int(11) unsigned NOT NULL default '0',
  `ticket_id` int(11) unsigned NOT NULL default '0',
  `staff_id` int(11) unsigned NOT NULL default '0',
  `user_id` int(11) unsigned not null default 0,
  `thread_type` enum('M','R','N') NOT NULL,
  `poster` varchar(128) NOT NULL default '',
  `source` varchar(32) NOT NULL default '',
  `title` varchar(255),
  `body` mediumtext NOT NULL,
  `format` varchar(16) NOT NULL default 'html',
  `ip_address` varchar(64) NOT NULL default '',
  `created` datetime NOT NULL,
  `updated` datetime NOT NULL,
  PRIMARY KEY  (`id`),
  KEY `ticket_id` (`ticket_id`),
  KEY `staff_id` (`staff_id`),
  KEY `pid` (`pid`)
) DEFAULT CHARSET=utf8;

**Comment by blastoma** [comment]
@ntozier but what about autor problems: "Email is truncated due to special chars" ?
I have the same problem, can it help me?

**Comment by greezybacon** [comment]
@ntozier That query is dangerous. It will drop all ticket threading history. To support the four-byte UTF-8 chars (**on MySql 5.6 / MariaDB 10.0 and higher**), use
```sql
ALTER TABLE `ost_thread_entry`
CHANGE `body` `body` MEDIUMTEXT NOT NULL CHARSET utf8mb4
```

**Comment by blastoma** [comment]
@greezybacon I have MariaDB 10.0.21, but mysql send error to me - http://joxi.ru/8AnGLOKuqwLBbr

**Comment by ntozier** [comment]
@greezybacon I agree that it could be dangerous, but in this specific case how is it dangerous?  He said that he doesn't have that table....

**Comment by greezybacon** [comment]
@ntozier the table was renamed between v1.9 and v1.10. That's why he doesn't have it. If he scanned his database and found the `ost_thread_entry` table (which it was renamed to), then he might regret adapting your code to that table.

**Comment by greezybacon** [comment]
@blastoma sorry for the confusion, I didn't test the statement. I updated my previous comment. Please try again.

**Comment by blastoma** [comment]
@greezybacon okey, I tried do it and now I have "????" instead smiles :)

**Comment by greezybacon** [comment]
I'm out of ideas. Perhaps upgrade to v1.9.12?

**Comment by ntozier** [comment]
This still sounds like a character or encoding issue on the db or table level.

in MySQL Workbench right click on the db name and select Schema inspector to see the default collation and default character sets.

Then you can also click on the tables tab.  Select the %thread table and right click on it.  Then you can use the table inspector and see the table collation.  Then you can click on the Columns and see the Character set.

**Comment by mfelber** [comment]
I have seen this behavior too (in 1.9.x and 1.10.x).
When I send an email to osticket with an emoticon (not as ":-)" but "=F0=9F=98=80") only the text before the emoticon is imported.

If the database column is "adjusted" with this statement from greezybacon the email will not be truncated anymore (!) but instead of the emoticon "????" is stored in the database (and shown in the thread entry).

When I try to insert the emoticon with mysql workbench via copy and paste it is shown as emoticon first but when I apply the insert it is also shown as "????".

After executing the statement "SET NAMES 'utf8mb4' COLLATE 'utf8mb4_unicode_ci'; with mysql workbench it is possible to store an emoticon via copy and paste (also with mysql workbench direct in the database). It is shown as emoticon in the database but osticket shows it as "?" in the thread entry.

When the mysql server is restarted the emoticon is also shown as "?" in the database until the "set names" statement is executed again. Then it is shown again as emoticon in the database.

utf8mb4 is not so easy to use like it sounds. It seems that there are changes in the php code and in the database environment necessary for it (maybe not only on the affected column?). Does anybody use utf8mb4 with osticket and are the emoticons shown in the thread entry like this "😀"?

I have not much knowledge about this things but it seems (due to the fact that utf8mb4 is able to store 4 bytes characters) that the maximum column length has to be adjusted and the performance is slower that utf8.

https://dev.mysql.com/doc/refman/5.5/en/charset-unicode-upgrading.html
https://mathiasbynens.be/notes/mysql-utf8mb4

Has anybody some experiences about this issues (adjusting column length for osticket, slower performance)?

**Comment by greezybacon** [comment]
@mfelber I took a stab at the issue in #3244.

**Comment by mfelber** [comment]
@greezybacon Looks terrific:

![terrific](https://cloud.githubusercontent.com/assets/5539456/16361628/df5fa980-3b96-11e6-8b4c-5f5e55b2138f.jpg)

It seems that adjusting the column and using the linked code is sufficient to get it work. I don't need to use the "set names" statement anymore...

**Comment by ponchofiesta** [comment]
Old issue but the problem is still present. I switched to utf8mb4 in MySQL. The text is not truncated anymore but the special char ist always shown as "?" now.
osTicket 1.10.1

This doesn't happen with Mails from Evolution. The mail header is encoded with utf-8. But if the header is encoded with some latin charset I get wrong chars.

(I'm not interrested in emojis but in other special chars like üöäß)

**Comment by ponchofiesta** [comment]
I assume the problem has something to do with the locale environment variables. I used a docker image based on Alpine Linux for osTicket. It doesn't have `locales` command available. The cron job is running on the shell and might have other environment variables than the PHP process.

I switched to a Ubuntu based docker image and set `export LANG=en_US.UTF-8`. Now The umlauts problem is gone.

**Comment by Advialance** [comment]
@ponchofiesta I just had the same problem with OsTicket v1.10.4
The problem is that OsTicket defaults the encoding to utf8_general_ci you will need to change encoding to 4 Byte utf8mb4_unicode_ci for some special characters to work properly (i.e. Emojis). To do so you have to do 2 things:
1. Convert your whole mysql db to utf8mb4_unicode_ci encoding if the database already exists you should change the databases collation as well as all tables collations and all columns collation. Especially the ones in ost_thread_entry
2. Make Osticket use the proper encoding in its connection. In 'include/mysqli.php' change
```php
//set desired encoding just in case mysql charset is not UTF-8 - Thanks to FreshMedia
    @db_set_all(array(  
        'NAMES'                 => 'utf8',  
        'CHARACTER SET'         => 'utf8',  
        'COLLATION_CONNECTION'  => 'utf8_general_ci',  
        'SQL_MODE'              => '',  
    ), 'session');  
    $__db->set_charset('utf8');
```  

to

```php
//set desired encoding just in case mysql charset is not UTF-8 - Thanks to FreshMedia
    @db_set_all(array(  
        'NAMES'                 => 'utf8mb4',  
        'CHARACTER SET'         => 'utf8mb4',  
        'COLLATION_CONNECTION'  => 'utf8mb4_unicode_ci',  
        'SQL_MODE'              => '',  
    ), 'session');  
    $__db->set_charset('utf8mb4');
```

**Comment by JediKev** [comment]
This has been resolved with #4523 until we switch to `utf8mb4` (with emoji support) slated for `v2.0` or later.

Cheers.
